packages/core/src/behavior.ts

```diff
@@ -0,0 +1,360 @@
+import {
+  EventObject,
+  InvokeCallback,
+  Subscribable,
+  Unsubscribable,
+  InterpreterOptions,
+  Spawnable,
+  Observer
+} from './types';
+import { ActorRef } from './Actor';
+import {
+  toSCXMLEvent,
+  isPromiseLike,
+  isObservable,
+  isMachineNode,
+  isSCXMLEvent
+} from './utils';
+import { doneInvoke, error, actionTypes } from './actions';
+import { isFunction } from 'util';
+import { MachineNode } from './MachineNode';
+import { interpret, Interpreter } from './interpreter';
+import { State } from './State';
+
+export interface ActorContext {
+  self: ActorRef<any, any>; // TODO: use type params
+  name: string;
+}
+
+export const startSignal = Symbol('xstate.start');
+export const stopSignal = Symbol('xstate.stop');
+
+export type LifecycleSignal = typeof startSignal | typeof stopSignal;
+
+/**
+ * An object that expresses the behavior of an actor in reaction to received events,
+ * as well as an optionally emitted stream of values.
+ *
+ * @template TReceived The received event
+ * @template TEmitted The emitted value
+ */
+export interface Behavior<TReceived extends EventObject, TEmitted> {
+  receive: (
+    actorContext: ActorContext,
+    event: TReceived
+  ) => Behavior<TReceived, TEmitted>;
+  receiveSignal: (
+    actorContext: ActorContext,
+    signal: LifecycleSignal
+  ) => Behavior<TReceived, TEmitted>;
+  /**
+   * The initial emitted value
+   */
+  initial: TEmitted;
+  subscribe?: (observer: Observer<TEmitted>) => Unsubscribable | undefined;
+}
+
+export function createCallbackBehavior<TEvent extends EventObject>(
+  callback: InvokeCallback,
+  parent?: ActorRef<any>
+): Behavior<TEvent, undefined> {
+  let canceled = false;
```

Comment:
isnt this against what behaviors should be? what i mean - from what i understand a behavior should describe how an actor should behave in a given "state" (which is not defined explicitly, but rather through behavior itself can be seen as state) and in general it should be possible to just restart actors by reusing initial behaviors, so it should merely encode the desired behavior, but should not contain a state of its own. This breaks this promise because this behavior has a mutable state in a closure

Reply:
Behaviors (even in Akka) can use closures. This can easily be represented as pure functions too - it's a trivial refactor.

Reply:
what i meant is that i was under the impression that behaviors should not use mutable state (they can use closures to capture some state, but they cant write to it), but maybe I was wrong - gonna check this out later, doesn't matter much for this PR

Reply:
I can implement this differently, sure.